File: algorithms-and-data-structures/binary_search.py
#!/usr/bin/env python3


def binary_search(element, a_list):
    """ Binary searches the given element within the specified sorted array.
    Complexity: O(log N)

    Args:
        element (int): Element to search for
        a_list (array[int]): Sorted array of integers to search within.

    Returns:
        index (int): If item is found, returns the index where the item was found, else, returns -1
    """
    head = 0
    tail = len(a_list) - 1
    pointer = (head + tail) // 2
    while head <= pointer <= tail:
        if a_list[pointer] == element:
            # Element found
            return pointer
        elif a_list[pointer] > element:
            tail = pointer - 1
        else:
            head = pointer + 1
        pointer = (head + tail) // 2
    return -1


def main():
    # Test case 1: Vanilla
    sample_list = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
    assert binary_search(0, sample_list) == 0
    assert binary_search(1, sample_list) == 1
    assert binary_search(2, sample_list) == 2
    assert binary_search(3, sample_list) == 3
    assert binary_search(4, sample_list) == 4
    assert binary_search(5, sample_list) == 5
    assert binary_search(6, sample_list) == 6
    assert binary_search(7, sample_list) == 7
    assert binary_search(8, sample_list) == 8
    assert binary_search(9, sample_list) == 9

    # Test case 2: search item is within range of the list, but not in list
    sample_list_1 = [1, 3, 5, 7, 9]
    sample_list_2 = [0, 2, 4, 6, 8]
    assert binary_search(2, sample_list_1) == -1
    assert binary_search(4, sample_list_1) == -1
    assert binary_search(6, sample_list_1) == -1
    assert binary_search(8, sample_list_1) == -1

    assert binary_search(1, sample_list_2) == -1
    assert binary_search(3, sample_list_2) == -1
    assert binary_search(5, sample_list_2) == -1
    assert binary_search(7, sample_list_2) == -1

    # Test case 3: search item is smaller than the smallest element in list
    sample_list_3 = [4, 5, 6, 7, 8]
    assert binary_search(-1, sample_list_3) == -1
    assert binary_search(2, sample_list_3) == -1

    # Test case 4: search item is larger than the largest element in list
    assert binary_search(10, sample_list_3) == -1
    assert binary_search(200, sample_list_3) == -1

    print("All test cases cleared")


if __name__ == '__main__':
    main()

File: algorithms-and-data-structures/BinarySearch/src/BinarySearch.java
public class BinarySearch {

    public static void main(String[] args) {
        int[] sampleList = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
        assert binarySearch(sampleList,0) == 0;
        assert binarySearch(sampleList,1) == 1;
        assert binarySearch(sampleList,2) == 2;
        assert binarySearch(sampleList,3) == 3;
        assert binarySearch(sampleList,4) == 4;
        assert binarySearch(sampleList,5) == 5;
        assert binarySearch(sampleList,6) == 6;
        assert binarySearch(sampleList,7) == 7;
        assert binarySearch(sampleList,8) == 8;
        assert binarySearch(sampleList,9) == 9;
        assert binarySearch(sampleList,-1) == -1;
        assert binarySearch(sampleList,10) == -1;

        int[] sampleList1 = {1, 3, 5, 7, 9};
        int[] sampleList2 = {0, 2, 4, 6, 8};
        assert binarySearch(sampleList1,2) == -1;
        assert binarySearch(sampleList1,4) == -1;
        assert binarySearch(sampleList1,6) == -1;
        assert binarySearch(sampleList1,8) == -1;

        assert binarySearch(sampleList2,1) == -1;
        assert binarySearch(sampleList2,3) == -1;
        assert binarySearch(sampleList2,5) == -1;
        assert binarySearch(sampleList2,7) == -1;

    }

    private static int binarySearch(int[] anArray, int element) {
        int head = 0;
        int tail = anArray.length - 1;
        int ptr = (head + tail) / 2;
        while (head <= tail) {
            if (anArray[ptr] == element) {
                return ptr;
            } else if (anArray[ptr] < element) {
                head = ptr + 1;
            } else {
                tail = ptr - 1;
            }
            ptr = (head + tail) / 2;
        }
        return -1;
    }
}
